lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,293 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'avro/io'
+require 'json'
+
+module Avro
+  module IO
+    # JsonDecoder keeps track of the data and reads it using a given data type.
+    class JsonDecoder
+      # memo-ized map of Schema::PrimitiveSchema objects so we're not allocating an object for every validation
+      PRIMITIVE_SCHEMAS = Hash[Schema::PRIMITIVE_TYPES_SYM.map { |k| [k, Schema::PrimitiveSchema.new(k)] }]
+
+      attr_accessor :data
+
+      def io_reader(reader)
+        datum = reader.read
+        datum = '""' if datum.empty?
+        data = JSON.parse(datum, {:quirks_mode => true})
+        @data = data
+
+        self
+      end
+
+      def read_null
+        validate(:null, data)
+        data
+      end
+
+      def read_boolean
+        validate(:boolean, data)
+        data
+      end
+
+      def read_int
+        validate(:int, data)
+        Integer(data)
+      end
+
+      def read_long
+        validate(:long, data)
+        Integer(data)
+      end
+
+      def read_float
+        validate(:float, data)
+        Float(data)
+      end
+
+      def read_double
+        read_float
+      end
+
+      def read_bytes
+        validate(:bytes, data)
+        data.encode(Encoding::ISO_8859_1, Encoding::UTF_8)
```

Comment:
Why does this change the encoding at all? Since bytes should be encoded as strings and non-printable characters (as defined by the JSON spec) should be encoded as "\uhhhh" I can understand some translation is needed here. But I don't think transcoding the string accomplishes that task.

If I try `"\\u0063".encode(ISO_8859_1, UTF_8)` in IRB, I don't get `"c"` as you would expect, so this doesn't decode escaped characters.